calling-apis/get-requests-with-axios.md

```diff
@@ -0,0 +1,453 @@
+
+<!-- Question 2 -->
+<!-- prettier-ignore-start -->
+### !challenge
+* type: multiple-choice
+* id: 7cbb2b75
+* title: GET Requests With axios
+##### !question
+
+In the case of handling an unsuccessful request, which of the following options gives back the HTTP response body, which contains the data that the API sent back?
+
+##### !end-question
+##### !options
+
+* `error`
+* `error.response.status`
+* `error.response.data`
+
+##### !end-options
+##### !answer
+
+* `error.response.data`
+
+##### !end-answer
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+### `finally`
+
+`axios` supports a `finally` clause, which will run after either the `then` or `catch` no matter what.
+
+The example syntax looks like this:
+
+```js
+const axios = require("axios");
+
+axios
+  .get("https://dog.ceo/api/breeds/list/all")
+  .then((response) => {
+    console.log("success!");
+  })
+  .catch((error) => {
+    console.log("error!");
+  })
+  .finally(() => {
+    console.log("this is always executed, no matter what!");
+  });
+```
+
+## Query Params
+
+As part of Jamel's DogSpotter app, he needs to find the latitude and longitude of Seattle, Washington, USA. He'll use the [LocationIQ Geocoding API](https://locationiq.com/sandbox/geocoding/forward).
+
+After looking at [the docs for Search/Forward Geocoding](https://locationiq.com/docs), Jamel determined that his request should look like this:
+
+- `GET` request to `https://us1.locationiq.com/v1/search.php`
+- Query Params:
+  - `key`: the API key
+  - `q`: A search term. In this case, `"Seattle, Washington, USA"`
+  - `format`: `"json"`
+
+He expects a response body that looks like:
+
+```json
+[
+  {
+    "place_id": "332439918643",
+    "licence": "https://locationiq.com/attribution",
+    "lat": "47.673455",
+    "lon": "-122.331537",
+    "display_name": "Seattle, King County, Washington, USA",
+    "boundingbox": ["47.495514", "47.734165", "-122.435956", "-122.236044"],
+    "importance": 0.25
+  }
+]
+```
+
+To add query params to our `GET` request, Jamel can research the `axios` documentation.
+
+He learns that to add query params, he utilizes `axios.get()`'s second parameter. This function takes in an optional "config" object.
+
+Within this "config" object, there we can add in a key `params`. The value of `params` should be another object.
+
+This `params` object should contain the key-value pairs we want to send as query params.
+
+Jamel's code eventually looks [like this](https://replit.com/@adacore/GET-Requests-With-axios-Demo#index.js):
+
+```js
+const axios = require("axios");
+
+axios
+  .get("https://us1.locationiq.com/v1/search.php", {
+    params: {
+      key: "... jamel's api key ...",
+      q: "Seattle, Washington, USA",
+      format: "json",
+    },
+  })
+  .then((response) => {
+    console.log("success!", response.data);
+  })
+  .catch((error) => {
+    console.log("error!", error.response.data);
+  });
+```
+
```

Comment:
Adds a section detailing how to set up the api key secret in replit.
```suggestion

Before trying to run this code ourselves, we should examine how Jamel configured his LocationIQ API key.

Recall that we should always try to keep our API keys private. During local development, we have created `.env` files to hold these values to help avoid accidentally committing them to Git. Other environments have provided management consoles where we can set secrets so that they are kept safe from prying eyes!

Replit.com provides a UI for creating secrets, which we can then access through the `process.env` object.

To set up our secret in our REPL, we click the Secrets icon (shaped like a padlock).

![The Replit.com interface showing the padlock-shaped icon highlighted. The hover text describes the icon as "Secrets (Environment variables)."](./../assets/calling-apis_get-requests-with-axios_secrets.png)  
_Fig. Locating the Secrets panel in Replit.com_

Then we fill in the key name with an appropriate name (we used `api_key`), paste our LocationIQ API key as the value, then click "Add new secret."

![The Replit.com interface showing the Secrets panel. The key field has "api_key" entered, and a placeholder in the value field of "YOUR_LOCATIONIQ_API_KEY" is displayed.](./../assets/calling-apis_get-requests-with-axios_secrets-api-key.png)  
_Fig. Filling in our secret API key_

Now our API key will be accessible in our code through `process.env['api_key']`!

```